Break out packages + substitutions docs

kbx81 opened the pull request:
## Description:

Breaks out package & substations docs into their own pages.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#<esphome PR number goes here>

## Checklist:

  - [ ] I am merging into `next` because this is new documentation that has a matching pull-request in [esphome](https://github.com/esphome/esphome) as linked above.  
    or
  - [x] I am merging into `current` because this is a fix, change and/or adjustment in the current documentation and is not for a new component or feature.

  - [x] Link added in `/index.rst` when creating new documents for new components or cookbook.

kbx81 committed:
Break out packages + substitutions docs

kbx81 committed:
Fix links

jesserockz commented on components/packages/index.rst:
```diff
@@ -0,0 +1,259 @@
+        # files: file1.yml
+
+      # shorthand form github://username/repository/[folder/]file-path.yml[@branch-or-tag]
+      remote_package_three: github://esphome/non-existant-repo/file1.yml@main
```
Lets flip this around so simpler comes first

jesserockz commented on components/packages/index.rst:
```diff
@@ -0,0 +1,259 @@
+      # A single file can be expressed using `file` or `files` as a string
+      remote_package_two:
+        url: https://github.com/esphome/non-existant-repo
+        file: file1.yml # cannot be combined with `files`
```
I regret adding `file` **and** `files` haha
I think we should undocument `file` and just use `files` everywhere

jesserockz commented on components/packages/index.rst:
```diff
@@ -0,0 +1,259 @@
+it into reusable pieces which can subsequently be used by many/all devices. Only unique pieces of your configuration
+remain in any given device's YAML configuration file.
+
+This can be accomplished with ESPHome's "packages" feature.
```
```suggestion
This can be accomplished with ESPHome's ``packages`` feature.
```

jesserockz commented on components/packages/index.rst:
```diff
@@ -0,0 +1,259 @@
+          door_location: right
+          open_switch_gpio: 15
+          close_switch_gpio: 18
+          open_duration: "1min"
+          close_duration: "50s"
+
+
+.. code-block:: yaml
+
+    # In garage-door.yaml
+    defaults:
+      open_duration: "2.1min"
+      close_duration: "2min"
+
+    switch:
+      - id: open_${door_location}_door_switch
+        name: ${door_name} Garage Door Open Switch
+        platform: gpio
+        pin: ${open_switch_gpio}
+
+      - id: close_${door_location}_door_switch
+        name: ${door_name} Garage Door Close Switch
+        platform: gpio
+        pin: ${close_switch_gpio}
+
+    cover:
+      - platform: time_based
+        name: ${door_name} Garage Door
+
+        open_action:
+          - switch.turn_on: open_${door_location}_door_switch
+        open_duration: ${open_duration}
+
+        close_action:
+          - switch.turn_on: close_${door_location}_door_switch
+        close_duration: ${close_duration}
+
+        stop_action:
+          - switch.turn_off: open_${door_location}_door_switch
+          - switch.turn_off: close_${door_location}_door_switch
```
Can we simplify these examples down to a bare minimum, just maybe a single entity with a single substitution and single default

jesserockz commented on components/packages/index.rst:
```diff
@@ -0,0 +1,259 @@
+Note how the piece of configuration describing ``api`` component in ``device_base.yaml`` gets merged with the services
+definitions from main configuration file.
+
+.. code-block:: yaml
```
Lets try to simplify this down. I don't have any specific suggestions though.

jesserockz commented on components/substitutions/index.rst:
```diff
@@ -0,0 +1,181 @@
+With substitutions, you can have a single generic source file for all nodes of one kind and
+substitute expressions in as required.
+
+.. code-block:: yaml
+
+    substitutions:
+      devicename: livingroom
+      upper_devicename: Livingroom
+
+    esphome:
+      name: $devicename
+      # ...
+
+    sensor:
+    - platform: dht
+      # ...
+      temperature:
+        name: ${upper_devicename} Temperature
+      humidity:
+        name: ${upper_devicename} Humidity
```
With the `friendly_name` being added, this is not a good example to have here.

nielsnl68 replied:
Are we still need to add this anyway? It is already prefix it.

jesserockz replied:
Yeah that was my point why this is a bad example. This example should do something else that is more useful to users.

jesserockz commented on components/packages/index.rst:
```diff
@@ -0,0 +1,259 @@
+
+.. _config-git_packages:
+
+Remote/Git Packages
```
Should have a `Configuration variables` section for this one as it has a specific schema.

jesserockz commented on components/substitutions/index.rst:
```diff
@@ -0,0 +1,181 @@
+
+.. _YAML-insertion-operator:
+
+YAML insertion operator
```
Not sure this fits in the `substitutions` docs.

kbx81 replied:
Agreed -- any thoughts on somewhere it _could_ go? I'm drawing a blank. But it is indeed useful...

jesserockz commented on components/substitutions/index.rst:
```diff
@@ -0,0 +1,181 @@
+      name: default
+      platform: ESP8266
+
+    esphome:
+      name: $name
+      platform: $platform
+      board: $board
```
We should not use the old, deprecated syntax here for an example.

kbx81 committed:
Formatting

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nielsnl68 commented on components/packages/index.rst:
```diff
@@ -0,0 +1,259 @@
+Packages as Templates
+---------------------
+
+Since packages are incorporated using the ``!include`` system, variables can be provided to them. This means that
```
When i checked this is the first time ``!include`` is mentioned in this document, would it be a good idea to explains what the differences are (when there are off-coerce).

nielsnl68 commented on components/packages/index.rst:
```diff
@@ -0,0 +1,259 @@
+.. code-block:: yaml
+
+    # In garage-door.yaml
+    defaults:
```
I use "substitutions" here what is the difference?

jesserockz replied:
the difference is that `defaults` are scoped to the package only. `substitutions` "leak" out to the root yaml merge. I think

nielsnl68 commented on components/packages/index.rst:
```diff
@@ -0,0 +1,259 @@
+
+For example, to set a specific update interval on a common uptime sensor that is shared between configurations:
+
+.. code-block:: yaml
+
+    packages:
+      common: !include common.yaml
+
+    sensor:
+      - id: !extend uptime_sensor
+        update_interval: 10s
```
This example does not mirror the documentation above.

nielsnl68 reviewed (commented):
just some points i was wondering about.

kbx81 committed:
Implement feedback from reviews

kbx81 commented:
@jesserockz @nielsnl68 I think I have addressed the feedback from your reviews in [`fcd27b9` (#4095)](https://github.com/esphome/esphome-docs/pull/4095/commits/fcd27b9991dcd7200bba019395e9bd5548d2a5c3). Please have a look and let me know -- thanks! 🍻

jesserockz commented on components/packages/index.rst:
```diff
@@ -0,0 +1,212 @@
+- **url** (**Required**, string): The URL for the repository.
+- **username** (*Optional*, string): Username to be used for authentication, if required.
+- **password** (*Optional*, string): Password to be used for authentication, if required.
+- **files** (*Optional*, list of strings): List of files to include. If omitted, all files in the repo are included.
```
Files is required

kbx81 replied:
`has_at_least_one_key`...right... 🤦🏻‍♂️

kbx81 committed:
Fix files

nagyrobi committed:
Merge branch 'current' into 20240729-packages-substitutions

jesserockz commented on components/packages/index.rst:
```diff
@@ -0,0 +1,212 @@
+    wifi:
+      ssid: !secret wifi_ssid
+      password: !secret wifi_password
+      domain: .yourdomain.lan
+      fast_connect: true
```
```suggestion
```

jesserockz commented on components/packages/index.rst:
```diff
@@ -0,0 +1,212 @@
+    # In device_base.yaml
+    esphome:
+      name: ${node_name}
+      platform: ESP32
+      board: wemos_d1_mini32
+      build_path: ./build/${node_name}
```
```suggestion

    esp32:
      board: wemos_d1_mini32
```

jesserockz commented on components/packages/index.rst:
```diff
@@ -0,0 +1,212 @@
+      scl: GPIOXX
+
+    logger:
+      level: ${log_level}
```
Not being supplied anywhere
```suggestion
```

jesserockz commented on components/packages/index.rst:
```diff
@@ -0,0 +1,212 @@
+    api:
+      encryption:
+        key: !secret api_encryption_key
+      reboot_timeout: 1h
```
```suggestion
```

jesserockz reviewed (changes requested):
Simplify even more =)
We only want people to understand that things can go in different files. We dont need to add so much detail, especially things that most people dont even add/change

kbx81 committed:
Apply suggestions from code review

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

kbx81 committed:
A little more simplification

kbx81 committed:
Merge branch 'current' into 20240729-packages-substitutions

kbx81 committed:
Merge branch 'current' into 20240729-packages-substitutions
